lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,366 @@
+               << "[checkStructure] " << contractionDims->m.size() << " "
+               << contractionDims->n.size() << " " << contractionDims->k.size()
+               << " " << contractionDims->batch.size() << "\n");
+    return failure();
+  }
+  unsigned classifiedLoops =
+      contractionDims->m.size() + contractionDims->n.size() +
+      contractionDims->k.size() + contractionDims->batch.size();
+  if (linalgOp.getNumLoops() != classifiedLoops) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "[checkStructure] Not all loops are classified\n");
+    return failure();
+  }
+
+  unsigned mAffinePos = contractionDims->m[0];
+  unsigned nAffinePos = contractionDims->n[0];
+  // contractionDims.k could be of 2 cases:
+  //     1. dims.k.size() == 2: non-VNNI, K = dims.k[1]
+  //     2. dims.k.size() == 3: VNNI, K = dims.k[1] * dims.k[2]
+  unsigned batchAffinePos = contractionDims->k.front();
+  SmallVector<unsigned, 2> kAffinePos;
+  if (contractionDims->k.size() == 2)
+    kAffinePos = {contractionDims->k[1]};
+  else if (contractionDims->k.size() == 3)
+    kAffinePos = {contractionDims->k[1], contractionDims->k[2]};
+  else
+    return failure();
+
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] Candidate dims: "
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] m pos in affine: " << mAffinePos
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] n pos in affine: " << nAffinePos
+                          << "\n");
+  for (auto kp : kAffinePos)
+    LLVM_DEBUG(llvm::dbgs()
+               << "[inferBrgemmDims] k pos in affine: " << kp << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] batch pos in affine: "
+                          << batchAffinePos << "\n");
+
+  OpOperand *operandA = linalgOp.getDpsInputOperands()[0];
+  OpOperand *operandB = linalgOp.getDpsInputOperands()[1];
+
+  BrgemmDims brgemmDims;
+
+  auto checkAndGetPosInDomain = [&](int64_t &dim, ArrayRef<unsigned> dimPos,
+                                    OpOperand *operand) {
+    auto pos = getPosInDomain(dimPos, operand, linalgOp);
+    assert(pos && "Cannot find position in codomain");
+    dim = *pos;
+  };
+
+  // A(batch, m, k)
+  checkAndGetPosInDomain(brgemmDims.batchDimA, batchAffinePos, operandA);
+  checkAndGetPosInDomain(brgemmDims.leadingDimA, {mAffinePos}, operandA);
+  checkAndGetPosInDomain(brgemmDims.minorDimA, kAffinePos, operandA);
+  // B(batch, k, n) or B(batch, k/vnni_step, n, vnni_step)
+  // note: B does not use VNNI format K affine
+  checkAndGetPosInDomain(brgemmDims.batchDimB, batchAffinePos, operandB);
+  checkAndGetPosInDomain(brgemmDims.leadingDimB, {kAffinePos[0]}, operandB);
+  checkAndGetPosInDomain(brgemmDims.minorDimB, {nAffinePos}, operandB);
+  // C(m, n)
+  // Currently useless, no need to set
+  // checkAndGetPosInDomain(brgemmDims.leadingDimC, {mAffinePos}, operandC);
+  // checkAndGetPosInDomain(brgemmDims.minorDimC, kAffinePos, operandC);
+
```

Comment:
Better to check if the dimA and dimB match or not? `batchDimA == batchDimB` etc.

Reply:
In this function we only calculate the position index of each dims, the actual value of dim size would be check in verification of `BrgemmOp`.